drivers/spi/spi_dw.c

```diff
@@ -439,12 +421,26 @@ static int transceive(const struct device *dev,
 		DW_SPI_IMR_UNMASK;
 	write_imr(info, reg_data);
 
-	spi_context_cs_control(&spi->ctx, true);
+	if (!spi_dw_is_slave(spi)) {
+		/* if cs is not defined as gpio, use hw cs */
+		if (spi_cs_is_gpio(config)) {
+			spi_context_cs_control(&spi->ctx, true);
+		} else {
+			write_ser(info, BIT(config->slave));
+		}
+	}
 
 	LOG_DBG("Enabling controller");
 	set_bit_ssienr(info);
 
 	ret = spi_context_wait_for_completion(&spi->ctx);
+
+#ifdef CONFIG_SPI_SLAVE
+	if (spi_context_is_slave(&spi->ctx) && !ret) {
+		ret = spi->ctx.recv_frames;
+	}
+#endif /* CONFIG_SPI_SLAVE */
+
```

Comment:
Could you split this commit in 2? one for the init macro, one for the slave fixes.

Reply:
The slave fixes are also within the init macro. It's kinda difficult to split them up

Reply:
done